### ISSUE TITLE
New update on Redux devtools breaks Jira site

### ISSUE THREAD

**Issue by BigsonLvrocha** [reduxjs/redux-devtools#1032]
With the new 3.0.4 version of the redux-devtools that went live yesterday the jira page won't load at all, just the side menu and some parts of the page, uninstalling the extension fixes the problem

I came here and saw no update was made, is there something wrong?

---
**Update by @timdorr:**
A fixed version is pending review with Google. In the meantime, I've posted some workarounds here: https://github.com/reduxjs/redux-devtools/issues/1032#issuecomment-1016837825

**Comment by Methuselah96** [comment]
Here's the update: https://github.com/reduxjs/redux-devtools/releases/tag/creat. Taking a look.

**Comment by bersamtrustyou** [comment]
I can confirm this.

**Comment by Methuselah96** [comment]
I'm having trouble reproducing the issue. I've loaded up a test Jira project and things seem to be working. What should I be trying to reproduce the issue?

![image](https://user-images.githubusercontent.com/693755/150152739-ac18a152-c879-4eec-ae3f-39610e295246.png)

**Comment by bersamtrustyou** [comment]
@Methuselah96 You're trying it on Jira Cloud which is Atlassian new product. In our case we are using Atlassian JIRA Project Management Software (v7.6.2) and exactly the same issue is happening as @BigsonLvrocha explained with 3.0.4.

**Comment by Methuselah96** [comment]
Alright, I'm publishing a revert, not sure how long it will take to go through (it takes anywhere from a couple hours to a day). If you need Redux DevTools in the meantime you can try out [Redux DevTools - Next](https://chrome.google.com/webstore/detail/redux-devtools-next/oamphgegmigmlgkdnijmeomjenbmkbdg).

**Comment by haskellcamargo** [comment]
I'm an Atlassian developer and I can confirm we've got more reports of this issue, and that disabling the Redux extension fixes the issue. Thanks for looking into this.

**Comment by Methuselah96** [comment]
@haskellcamargo Is there a way for me to test this out so I can figure out what's going on? I'm having trouble figuring out how to access the non-cloud version of Jira.

**Comment by haskellcamargo** [comment]
All the production tenants I've provisioned seem to be unaffected. I'm only able to reproduce the issue in our internal staging network. I'll post here some details we know about the problem:
- The first error happens in a file with the following expression: `this.vent.on = ...`. With Redux Devtools disabled, `this` has the context of the function. With it enabled, `this` refers to the global `window` object.
- I suspect Redux is proxying some values and improperly setting the value of `this`

A snippet of the offending code was found by inspecting the source tab in the Chrome devtools. The code is part of Marionette JS 2.1.0. We might be able to find the original source in the Backbone org.

```js
g = function () {
    this._channels = {};
    this.vent = {};
    this.commands = {};
    this.reqres = {};
    this._proxyMethods();
};
_.extend(g.prototype, {
    channel: function (name) {
        if (!name) throw Error('Channel must receive a name');
        return this._getChannel(name);
    },
    _getChannel: function (name) {
        var channel = this._channels[name];
        channel || ((channel = new h.Channel(name)), (this._channels[name] = channel));
        return channel;
    },
    _proxyMethods: function () {
        // In this method, `this` refers to `window` when Redux Devtools is enabled. It refers to `g` otherwise.
        _.each(
            ['vent', 'commands', 'reqres'],
            function (a) {
                _.each(
                    ['on', 'off', 'trigger', 'once', 'stopListening', 'listenTo', 'listenToOnce'],
                    function (b) {
                        var e = this;
                        // This is the line that fails due to the issue above:
                        this[a][b] = function (c) {
                            var d = e._getChannel(c)[a],
                                f = Array.prototype.slice.call(arguments, 1);
                            return d[b].apply(d, f);
                        };
                    },
                    this,
                );
            },
            this,
        );
    },
});
```

**Comment by camoore1979** [comment]
I can confirm that the extension breaks Jira (standard cloud). Specifically, my ability to access Kanban and Scrum boards. I'm able to view and edit tickets but get the following when I go to our scrum board. Disabling the Redux Dev Tools extension fixes the problem & everything works.

<img width="1778" alt="Screen Shot 2022-01-19 at 9 32 31 AM" src="https://user-images.githubusercontent.com/4583280/150166309-a860f0d8-b596-4a2f-93a3-d92bfa89ea65.png">
.

**Comment by Methuselah96** [comment]
Thanks for the helpful info. I'm able to reproduce the `Uncaught TypeError: Cannot read properties of undefined (setting 'on')` error it just doesn't break the view for me. Digging into it.

**Comment by haskellcamargo** [comment]
@Methuselah96 no worries! What screen are you on? Do you mind pasting the relative path to it? These files get loaded on pretty much every page, but they're absolutely necessary only for some of them, such as classic backlogs and boards (company managed). It might be the case you're testing on a next-gen board (also known as team managed).

**Comment by Methuselah96** [comment]
I'm on `jira/software/projects/TEST/boards/1`. I just created a test Jira board today to try to figure out what was going on.

**Comment by haskellcamargo** [comment]
@Methuselah96 that is a next-gen project, which should be unaffected. Try going to the **Projects** tab and creating a project that is **Company-managed**. It can be either Scrum or Kanban.

**Comment by haskellcamargo** [comment]
Awesome! Thanks @Methuselah96.

**Comment by haskellcamargo** [comment]
@Methuselah96 by the way, have you published the revert in the meantime? I need to send comms to the team and customers so they know what to expect. Thanks again.

**Comment by Methuselah96** [comment]
Figured out the issue. Redux was importing all of Lodash in https://github.com/reduxjs/redux-devtools/pull/1031 which was setting `window._`. Jira was trying to use `window._` as the underscore library.

**Comment by antonmordik** [comment]
As a temporary quick solution for chrome you can enable it only for a specific url

![image](https://user-images.githubusercontent.com/31385311/150177928-ac4dec1c-97b5-49ca-86f2-88f3067065ce.png)

**Comment by timdorr** [comment]
> Figured out the issue. Redux was importing all of Lodash in #1031 which was setting window._. Jira was trying to use window._ as the underscore library.

We removed lodash in Redux a while ago, though for different reasons. Let me know if you want help reimplementing the functions used there, or if you just want to refactor the whole `persistState` function.

**Comment by Methuselah96** [comment]
Thanks, importing the Lodash from the function entry-points fixes the issue (i.e., `import mapValues from 'lodash/mapValues';` instead of `import _ from 'lodsah';`) so I should be good for now.

**Comment by bbsdddougla** [comment]
It appears that this same issue broke part of another site that uses lodash: mavenlink.com. Specifically: Mavenlink's timesheets control does not load with the error `TypeError: _.pluck is not a function`.

I'll reevaluate after installing the updated plugin.

**Comment by timdorr** [comment]
BTW, for folks dealing with this, we are currently waiting on Google to approve the fixed version:

![image](https://user-images.githubusercontent.com/17882/150206916-e6975c7c-2845-4346-8152-7075828c9192.png)

If you need to fix this right now, you've got a few options:

- Disable or uninstall the extension temporarily
- Go to the Chrome Extension settings page and hit the Details button next to Redux DevTools (`chrome://extensions/?id=lmhkpmbekcpmknklioeibfkpmmfibljd`). Under "Site access", change the dropdown to "On specific sites" and add the domains/URLs where you want the extension to operate.
  
  ![image](https://user-images.githubusercontent.com/17882/150207548-336f03b2-dcdb-49d8-af08-e9217afb0a93.png)
- Install the "Next" version of the extension, which is the version prior to this change: https://chrome.google.com/webstore/detail/redux-devtools-next/oamphgegmigmlgkdnijmeomjenbmkbdg

I'll update this thread as soon as the fixed version is published. 

Also, the Firefox version is still at 3.0.3, so it's unaffected by this. We'll get them in sync once we've confirmed the Chrome version is working.

**Comment by timdorr** [comment]
Just got word that 3.0.5 is published, so it should be fixed once you get the update. You can hit the Update button on the Extensions page to force it. 

I'm going to close this out, but feel free to comment if this doesn't fix things. We can investigate if there is a problem with the update or if there is a separate issue.

**Comment by haskellcamargo** [comment]
Awesome! Great work. I've tested it out and it's working again.
We'll update the [status page for Jira](https://jira-software.status.atlassian.com/) with the great news!

**Comment by Methuselah96** [comment]
I wrote up a [postmortem](https://nathanbierema.com/redux-devtools-jira/) for those interested.
